subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,252 @@
+
+static bool mtrace_active;
+
+/*
+ * SRAM window for mtrace is organized in 16 (equal sized) pages:
+ *
+ *	------------------------
+ *	| Page0  - descriptors |
+ *	------------------------
+ *	| Page1  - slot0       |
+ *	------------------------
+ *	| Page2  - slot1       |
+ *	------------------------
+ *	|         ...          |
+ *	------------------------
+ *	| Page14  - slot13     |
+ *	------------------------
+ *	| Page15  - slot14     |
+ *	------------------------
+ *
+ * The slot size == page size
+ *
+ * The first page contains descriptors for the remaining 15 cores
+ * The slot descriptor is:
+ * u32 res_id;
+ * u32 type;
+ * u32 vma;
+ *
+ * Log buffer slots have the following layout:
+ * u32 host_read_ptr;
+ * u32 dsp_write_ptr;
+ * u8 buffer[];
+ *
+ * The two pointers are offsets within the buffer.
+ */
+
+#define MTRACE_DESCRIPTOR_SIZE		12 /* 3 x u32 */
+#define MTRACE_PAGE_SIZE		0x1000
+#define MTRACE_SLOT_SIZE		MTRACE_PAGE_SIZE
+
+/* debug log slot types */
+#define MTRACE_SLOT_UNUSED			0x00000000
+#define MTRACE_SLOT_CRITICAL_LOG		0x54524300 /* byte 0: core ID */
+#define MTRACE_SLOT_DEBUG_LOG		0x474f4c00 /* byte 0: core ID */
+#define MTRACE_SLOT_GDB_STUB		0x42444700
+#define MTRACE_SLOT_TELEMETRY		0x4c455400
+#define MTRACE_SLOT_BROKEN			0x44414544
+ /* for debug and critical types */
+#define MTRACE_SLOT_CORE_MASK		GENMASK(7, 0)
+#define MTRACE_SLOT_TYPE_MASK		GENMASK(31, 8)
+
+#define MTRACE_LOG_BUF_SIZE		(MTRACE_SLOT_SIZE - 2 * sizeof(uint32_t))
+
+#define MTRACE_LOGGING_SLOT_TYPE(n)	(MTRACE_SLOT_DEBUG_LOG | (n & MTRACE_SLOT_CORE_MASK))
+
+#define MTRACE_DESCRIPTOR(n) ((uint8_t *)(z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE) + n * MTRACE_DESCRIPTOR_SIZE))
+
+#define MTRACE_LOG_BUF(n) (((uint8_t *)z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE)) + (n + 1) * MTRACE_SLOT_SIZE)
+
+static mem_addr_t mtrace_desc_type(int n)
+{
+	uint8_t *base = z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE);
+	uint8_t *desc_base = base + n * MTRACE_DESCRIPTOR_SIZE;
+
+	return (mem_addr_t)(desc_base + 4);
+}
+
+static void mtrace_init(void)
+{
+        uint32_t desc_type_slot0 = sys_read32(mtrace_desc_type(0));
+
+	if (desc_type_slot0 == MTRACE_LOGGING_SLOT_TYPE(0))
+		return;
+
+	sys_write32(MTRACE_LOGGING_SLOT_TYPE(0), mtrace_desc_type(0));
+}
+
+static void mtrace_out(int8_t *str, size_t len, size_t *space_left)
+{
+	uint8_t *base = MTRACE_LOG_BUF(0);
+	uint8_t *data = base + 8;
+	uint32_t r = sys_read32((mem_addr_t)base);
+	uint32_t w = sys_read32((mem_addr_t)(base + 4));
+	size_t avail, left;
+
+	/* FIXME: the WIN2 area gets overwritten/zeroed
+	 *        some time after soc_trace_init(), so as
+	 *        a stopgap, keep reinitializing the WIN2
+	 *        headers */
+	mtrace_init();
+
+	if (w > r)
+		avail = MTRACE_LOG_BUF_SIZE - w + r - 1;
+	else if (w == r)
+		avail = MTRACE_LOG_BUF_SIZE - 1;
+	else
+		avail = r - w - 1;
+
+
+	if (len == 0)
+		goto out;
+
+	/* data that doesn't fit is dropped */
+	left = MIN(avail, len);
```

Comment:
Relevant to the earlier not: this is a spot where having a single 60k buffer would be a lot less subject to overflow than 15 4k ones.  Even something like dumping a medium-scale stack trace can hit 4k.

Reply:
Ack, this is definitely a limitation. Use of dictionaries will help somewhat. Also, with the IPC notifications, the host will immediately get an IPC IRQ when less than 2k is available, so it probably can start consuming sooner than it does with winstream/ADSP backend.